File: src/lib/calendar/EventList.ts
/**
 * Wrapper estilo ArrayList<CalendarEvent>.
 *
 * Conceitos AP CS A:
 *  - Unit 7 (ArrayList): add, remove, set, size, get — todos os métodos
 *    aparecem no CED e na prova. Aqui implementamos a interface usando
 *    um array nativo por baixo.
 *  - Unit 4 (Iteration): traverse for-each para filtrar (filterByDate)
 *    e para acumular (countByCategory).
 *  - Unit 7 (Sorting): selection sort O(n²) escrita "à mão" para casar
 *    com o que cai no exame, embora pudéssemos chamar Array#sort.
 *
 * Por que não usar Array nativo direto? Porque a ideia é mostrar a API
 * que o aluno vê no exame, e todo o resto do código consome essa API.
 */

import { CalendarEvent, type EventCategory } from "./CalendarEvent"

export class EventList {
  // Unit 7: a ArrayList por baixo dos panos é um array dinâmico.
  private items: CalendarEvent[]

  public constructor(initial: CalendarEvent[] = []) {
    this.items = [...initial] // defensive copy
  }

  // ───────── métodos da API ArrayList do AP CED ─────────

  public size(): number {
    return this.items.length
  }

  public isEmpty(): boolean {
    return this.items.length === 0
  }

  public get(index: number): CalendarEvent {
    if (index < 0 || index >= this.items.length) {
      throw new RangeError(`EventList.get: índice ${index} fora dos limites`)
    }
    return this.items[index]
  }

  public add(event: CalendarEvent): boolean {
    this.items.push(event)
    return true
  }

  public set(index: number, event: CalendarEvent): CalendarEvent {
    if (index < 0 || index >= this.items.length) {
      throw new RangeError(`EventList.set: índice ${index} fora dos limites`)
    }
    const old = this.items[index]
    this.items[index] = event
    return old
  }

  public remove(index: number): CalendarEvent {
    if (index < 0 || index >= this.items.length) {
      throw new RangeError(
        `EventList.remove: índice ${index} fora dos limites`,
      )
    }
    return this.items.splice(index, 1)[0]
  }

  // Unit 4 + 7: linear search por id.
  public removeById(id: string): boolean {
    for (let i = 0; i < this.items.length; i++) {
      if (this.items[i].getId() === id) {
        this.items.splice(i, 1)
        return true
      }
    }
    return false
  }

  public indexOfId(id: string): number {
    for (let i = 0; i < this.items.length; i++) {
      if (this.items[i].getId() === id) return i
    }
    return -1
  }

  // ───────── filtros & utilitários (consumidos pela UI) ─────────

  // Unit 4: traverse + cópia condicional.
  public filterByDate(year: number, month: number, day: number): EventList {
    const out = new EventList()
    for (const ev of this.items) {
      if (ev.isOnDate(year, month, day)) out.add(ev)
    }
    return out
  }

  // Unit 4: traverse + acumulador. Devolve dicionário por categoria.
  public countByCategory(): Record<EventCategory, number> {
    const counts: Record<EventCategory, number> = {
      study: 0,
      exam: 0,
      assignment: 0,
      school_day: 0,
      special_day: 0,
    }
    for (const ev of this.items) {
      counts[ev.getCategory()]++
    }
    return counts
  }

  // Devolve o array nativo (cópia) — usado na UI quando React precisa.
  public toArray(): CalendarEvent[] {
    return [...this.items]
  }

  // Unit 7: Selection Sort por data (in place).
  // Algoritmo do CED: a cada passo, encontra o menor do "resto" e troca
  // com a posição corrente.
  public selectionSortByDate(): void {
    const n = this.items.length
    for (let i = 0; i < n - 1; i++) {
      let minIdx = i
      for (let j = i + 1; j < n; j++) {
        if (this.items[j].compareDate(this.items[minIdx]) < 0) {
          minIdx = j
        }
      }
      if (minIdx !== i) {
        // swap
        const tmp = this.items[i]
        this.items[i] = this.items[minIdx]
        this.items[minIdx] = tmp
      }
    }
  }

  // Versão imutável para quando a UI quer uma lista ordenada sem
  // mexer na original.
  public sortedByDate(): CalendarEvent[] {
    const copy = new EventList(this.items)
    copy.selectionSortByDate()
    return copy.toArray()
  }

  // ───────── filtros e busca ─────────

  // Unit 4: filtra por categoria.
  public filterByCategory(category: EventCategory): EventList {
    const out = new EventList()
    for (const ev of this.items) {
      if (ev.getCategory() === category) out.add(ev)
    }
    return out
  }

  // Unit 4: linear search por título (case-insensitive, substring).
  public linearFilterByTitle(query: string): CalendarEvent[] {
    const q = query.trim().toLowerCase()
    if (q.length === 0) return []
    const out: CalendarEvent[] = []
    for (const ev of this.items) {
      if (ev.getTitle().toLowerCase().includes(q)) out.push(ev)
    }
    return out
  }

  // ───────── ordenação por título (Unit 7) ─────────

  // Unit 7: Insertion Sort por título (in place). Pré-requisito do BS.
  public insertionSortByTitle(): void {
    const n = this.items.length
    for (let i = 1; i < n; i++) {
      const key = this.items[i]
      let j = i - 1
      while (j >= 0 && this.items[j].compareTitle(key) > 0) {
        this.items[j + 1] = this.items[j]
        j--
      }
      this.items[j + 1] = key
    }
  }

  public sortedByTitle(): CalendarEvent[] {
    const copy = new EventList(this.items)
    copy.insertionSortByTitle()
    return copy.toArray()
  }

  // Unit 7: Binary Search por título (exact match, case-insensitive).
  // Pré-condição: lista ORDENADA por título. Retorna índice ou -1.
  public binarySearchByTitle(title: string): number {
    const target = title.trim().toLowerCase()
    let lo = 0
    let hi = this.items.length - 1
    while (lo <= hi) {
      const mid = Math.floor((lo + hi) / 2)
      const cmp = this.items[mid]
        .getTitle()
        .toLowerCase()
        .localeCompare(target, "pt-BR")
      if (cmp === 0) return mid
      if (cmp < 0) lo = mid + 1
      else hi = mid - 1
    }
    return -1
  }

  // ───────── detector de conflito de horário (Unit 4) ─────────

  // Dois intervalos [aStart, aEnd) e [bStart, bEnd) se sobrepõem se
  // e somente se aStart < bEnd && bStart < aEnd.
  public findOverlapping(
    start: Date,
    durationMinutes: number,
    excludeId?: string,
  ): CalendarEvent[] {
    const out: CalendarEvent[] = []
    const aStart = start.getTime()
    const aEnd = aStart + Math.max(0, durationMinutes) * 60_000
    for (const ev of this.items) {
      if (excludeId && ev.getId() === excludeId) continue
      if (ev.getDurationMinutes() <= 0) continue
      const bStart = ev.getDate().getTime()
      const bEnd = bStart + ev.getDurationMinutes() * 60_000
      if (aStart < bEnd && bStart < aEnd) out.push(ev)
    }
    return out
  }
}

File: java/EventList.java
/*
 * EventList — wrapper estilo ArrayList<CalendarEvent>.
 *
 * Unit 7 — ArrayList API completa (add, get, set, remove, size).
 * Unit 7 — Selection Sort (por data) + Insertion Sort (por titulo).
 * Unit 7 — Binary Search por titulo (O(log n)).
 * Unit 4 — Linear search + filtros (filterByCategory, filterByDate).
 * Unit 4 — findOverlapping: deteccao de conflito de intervalos.
 */
import java.util.ArrayList;

public class EventList {

    // Por baixo dos panos, e um ArrayList. A classe so adiciona
    // metodos especificos do dominio.
    private ArrayList<CalendarEvent> items;

    public EventList() {
        this.items = new ArrayList<CalendarEvent>();
    }

    // ---------- API basica do ArrayList (Unit 7) ----------

    public int size() {
        return items.size();
    }

    public boolean isEmpty() {
        return items.size() == 0;
    }

    public CalendarEvent get(int index) {
        return items.get(index);
    }

    public boolean add(CalendarEvent ev) {
        return items.add(ev);
    }

    public CalendarEvent set(int index, CalendarEvent ev) {
        return items.set(index, ev);
    }

    public CalendarEvent remove(int index) {
        return items.remove(index);
    }

    // ---------- Linear search (Unit 4) ----------

    public boolean removeById(int id) {
        for (int i = 0; i < items.size(); i++) {
            if (items.get(i).getId() == id) {
                items.remove(i);
                return true;
            }
        }
        return false;
    }

    // ---------- Filtros (Unit 4: traverse + copia condicional) ----------

    public EventList filterByCategory(String category) {
        EventList out = new EventList();
        for (int i = 0; i < items.size(); i++) {
            CalendarEvent ev = items.get(i);
            if (ev.getCategory().equals(category)) {
                out.add(ev);
            }
        }
        return out;
    }

    public EventList filterByDate(int year, int month, int day) {
        EventList out = new EventList();
        for (int i = 0; i < items.size(); i++) {
            CalendarEvent ev = items.get(i);
            if (ev.isOnDate(year, month, day)) {
                out.add(ev);
            }
        }
        return out;
    }

    // Substring search no titulo (case-insensitive).
    public ArrayList<CalendarEvent> linearFilterByTitle(String query) {
        ArrayList<CalendarEvent> out = new ArrayList<CalendarEvent>();
        String q = query.trim().toLowerCase();
        if (q.length() == 0) return out;
        for (int i = 0; i < items.size(); i++) {
            CalendarEvent ev = items.get(i);
            if (ev.getTitle().toLowerCase().indexOf(q) >= 0) {
                out.add(ev);
            }
        }
        return out;
    }

    // ---------- Selection Sort por data (Unit 7) ----------
    //
    // Algoritmo classico do CED: a cada iteracao, acha o MENOR do
    // resto e troca com a posicao corrente.
    public void selectionSortByDate() {
        int n = items.size();
        for (int i = 0; i < n - 1; i++) {
            int minIdx = i;
            for (int j = i + 1; j < n; j++) {
                if (items.get(j).compareDate(items.get(minIdx)) < 0) {
                    minIdx = j;
                }
            }
            if (minIdx != i) {
                // swap
                CalendarEvent tmp = items.get(i);
                items.set(i, items.get(minIdx));
                items.set(minIdx, tmp);
            }
        }
    }

    // ---------- Insertion Sort por titulo (Unit 7) ----------
    //
    // Pre-requisito do binary search abaixo: BS exige lista ordenada.
    public void insertionSortByTitle() {
        int n = items.size();
        for (int i = 1; i < n; i++) {
            CalendarEvent key = items.get(i);
            int j = i - 1;
            while (j >= 0 && items.get(j).compareTitle(key) > 0) {
                items.set(j + 1, items.get(j));
                j--;
            }
            items.set(j + 1, key);
        }
    }

    // ---------- Binary Search por titulo (Unit 7) ----------
    //
    // Pre-condicao: lista ORDENADA por titulo (rodar insertionSortByTitle
    // antes). Retorna o indice do match exato, ou -1.
    // Complexidade: O(log n).
    public int binarySearchByTitle(String title) {
        String target = title.trim().toLowerCase();
        int lo = 0;
        int hi = items.size() - 1;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            String midTitle = items.get(mid).getTitle().toLowerCase();
            int cmp = midTitle.compareTo(target);
            if (cmp == 0) return mid;
            if (cmp < 0) lo = mid + 1;
            else         hi = mid - 1;
        }
        return -1;
    }

    // ---------- Detector de conflito de horario (Unit 4) ----------
    //
    // Dois intervalos [aStart, aEnd) e [bStart, bEnd) se sobrepoem
    // se e somente se aStart < bEnd && bStart < aEnd.
    public ArrayList<CalendarEvent> findOverlapping(
            int year, int month, int day, int hour, int minute,
            int durationMinutes) {
        ArrayList<CalendarEvent> out = new ArrayList<CalendarEvent>();
        // Converte o intervalo de entrada em minutos absolutos
        int aStart = totalMinutes(year, month, day, hour, minute);
        int aEnd = aStart + durationMinutes;
        for (int i = 0; i < items.size(); i++) {
            CalendarEvent ev = items.get(i);
            int dur = ev.getDurationMinutes();
            if (dur <= 0) continue;  // assignments nunca conflitam
            int bStart = totalMinutes(
                ev.getYear(), ev.getMonth(), ev.getDay(),
                ev.getHour(), ev.getMinute()
            );
            int bEnd = bStart + dur;
            if (aStart < bEnd && bStart < aEnd) {
                out.add(ev);
            }
        }
        return out;
    }

    // Helper: converte data/hora em minutos absolutos (mesma logica
    // de CalendarEvent.toMinutes, mas estatica).
    private static int totalMinutes(int y, int m, int d, int h, int mi) {
        int days = (y * 12 + m) * 31 + d;
        return days * 24 * 60 + h * 60 + mi;
    }

    // Acesso ao array interno (read-only — cuidar de nao mutar).
    public ArrayList<CalendarEvent> toArrayList() {
        return new ArrayList<CalendarEvent>(items);
    }
}
